Objects/dictobject.c

```diff
@@ -2510,6 +2562,28 @@ PyDict_Copy(PyObject *o)
             _PyObject_GC_TRACK(split_copy);
         return (PyObject *)split_copy;
     }
+
+    if (PyDict_CheckExact(mp) &&
+            mp->ma_values == NULL && mp->ma_keys->dk_refcnt == 1 &&
+            mp->ma_used >= mp->ma_keys->dk_nentries - 3)
+    {
+        /* Use fast-copy if:
+
+           (1) 'mp' is an instance of a subclassed dict; and
+
+           (2) 'mp' is not a split-dict; and
+
+           (3) if 'mp' is non-compact ('del' operation does not resize dicts),
+               do fast-copy only if it has at most 3 non-used keys.
```

Comment:
Instead of an absolute number, why not using a ratio? 3 unused is may a lot if the dict contains 16 items, but what is the dict contains 2^16 items?

Would it make sense to use a ratio, tolerate a maximum of 10% unused entries? Or 1%?

What does impact performances of dict.set(): the absolute number or the ratio?

If it's the absolute number, just explain that above, something like: "each unused entry make dict.set() slower, so create a compacted dict for optimal performances".